src/field_theory/finite.lean

```diff
@@ -118,22 +140,81 @@ begin
+  exact nat.pow_dvd_pow _ n.2,
 end
 
-lemma pow_card_sub_one_eq_one [field K] [fintype K] (a : K) (ha : a ≠ 0) :
-  a ^ (fintype.card K - 1) = 1 :=
-calc a ^ (fintype.card K - 1) = (units.mk0 a ha ^ (fintype.card K - 1) : units K) :
-    by rw [units.coe_pow, units.coe_mk0]
-  ... = 1 : by { classical, rw [← card_units, pow_card_eq_one], refl }
+lemma forall_pow_eq_one_iff (i : ℕ) :
+  (∀ x : units K, x ^ i = 1) ↔ q - 1 ∣ i :=
+begin
+  obtain ⟨x, hx⟩ := is_cyclic.exists_generator (units K),
+  classical,
+  rw [← card_units, ← order_of_eq_card_of_forall_mem_gpowers hx, order_of_dvd_iff_pow_eq_one],
+  split,
+  { intro h, apply h },
+  { intros h y,
+    rw ← powers_eq_gpowers at hx,
+    rcases hx y with ⟨j, rfl⟩,
+    rw [← pow_mul, mul_comm, pow_mul, h, one_pow], }
+end
+
+/-- The sum of `x ^ i` as `x` ranges over the units of a finite field of cardinality `q`
+is equal to `0` unless `(q - 1) ∣ i`, in which case the sum is `q - 1`. -/
+lemma sum_pow_units (i : ℕ) :
+  ∑ x : units K, (x ^ i : K) = if (q - 1) ∣ i then -1 else 0 :=
+begin
+  let φ : units K →* K :=
+  { to_fun   := λ x, x ^ i,
+    map_one' := by rw [units.coe_one, one_pow],
+    map_mul' := by { intros, rw [units.coe_mul, mul_pow] } },
+  haveI : decidable (φ = 1) := by { classical, apply_instance },
+  calc ∑ x : units K, φ x = if φ = 1 then fintype.card (units K) else 0 : sum_hom_units φ
+                      ... = if (q - 1) ∣ i then -1 else 0 : _,
+  suffices : (q - 1) ∣ i ↔ φ = 1,
+  { simp only [this],
+    split_ifs with h h, swap, refl,
+    rw [card_units, nat.cast_sub, cast_card_eq_zero, nat.cast_one, zero_sub],
+    show 1 ≤ q, from fintype.card_pos_iff.mpr ⟨0⟩ },
+  rw [← forall_pow_eq_one_iff, monoid_hom.ext_iff],
+  apply forall_congr, intro x,
+  rw [units.ext_iff, units.coe_pow, units.coe_one, monoid_hom.one_apply],
+  refl,
+end
+
+/-- The sum of `x ^ i` as `x` ranges over a finite field of cardinality `q`
+is equal to `0` if `i < q - 1`. -/
+lemma sum_pow_lt_card_sub_one (i : ℕ) (h : i < q - 1) :
+  ∑ x : K, x ^ i = 0 :=
+begin
+  by_cases hi : i = 0,
+  { simp only [hi, add_monoid.smul_one, sum_const, pow_zero, card_univ, cast_card_eq_zero], },
+  classical,
+  have hiq : ¬ (q - 1) ∣ i, { contrapose! h,  exact nat.le_of_dvd (nat.pos_of_ne_zero hi) h },
+  let φ : units K ↪ K := ⟨coe, units.ext⟩,
+  have : univ.map φ = univ \ finset.singleton 0,
+  { ext x,
+    simp only [true_and, embedding.coe_fn_mk, mem_sdiff, units.exists_iff_ne_zero,
+               mem_univ, mem_map, exists_prop_of_true, mem_singleton] },
+  calc ∑ x : K, x ^ i = ∑ x in univ \ finset.singleton 0, x ^ i :
+    by rw [← sum_sdiff (subset_univ (finset.singleton (0:K))), sum_singleton,
+           zero_pow (nat.pos_of_ne_zero hi), add_zero]
+    ... = ∑ x : units K, x ^ i : by { rw [← this, univ.sum_map φ], refl }
+    ... = 0 : by { rw [sum_pow_units K i, if_neg], exact hiq, }
+end
 
 end finite_field
 
 namespace zmod
 
-open finite_field
+open finite_field polynomial
```

Comment:
Are these actually necessary? It seems unlikely. This and `open nat` below.

Reply:
Reduced the number of `open`s in the file.